# Q1).Implement a KNN model to classify the animals in to categorie.

In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
df = pd.read_csv('Zoo.csv')
df.head()

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0    aardvark     1         0     0     1         0        0         1   
1    antelope     1         0     0     1         0        0         0   
2        bass     0         0     1     0         0        1         1   
3        bear     1         0     0     1         0        0         1   
4        boar     1         0     0     1         0        0         1   

   toothed  backbone  breathes  venomous  fins  legs  tail  domestic  catsize  \
0        1         1         1         0     0     4     0         0        1   
1        1         1         1         0     0     4     1         0        1   
2        1         1         0         0     1     0     1         0        0   
3        1         1         1         0     0     4     0         0        1   
4        1         1         1         0     0     4     1         0        1   

   type  
0     1  
1     1  
2     4  
3     1  
4     1

In [18]:
df.corr()

animal name      hair  feathers      eggs      milk  airborne  \
animal name     1.000000 -0.109112  0.058483  0.113466 -0.138181  0.048277   
hair           -0.109112  1.000000 -0.427851 -0.817382  0.878503 -0.198431   
feathers        0.058483 -0.427851  1.000000  0.419248 -0.410761  0.656553   
eggs            0.113466 -0.817382  0.419248  1.000000 -0.938848  0.376646   
milk           -0.138181  0.878503 -0.410761 -0.938848  1.000000 -0.366765   
airborne        0.048277 -0.198431  0.656553  0.376646 -0.366765  1.000000   
aquatic         0.023475 -0.473554 -0.058552  0.376244 -0.362613 -0.172638   
predator       -0.027931 -0.154769 -0.104430  0.011605 -0.029721 -0.295181   
toothed        -0.130423  0.492531 -0.613631 -0.642150  0.628168 -0.594311   
backbone       -0.058669  0.191681  0.231403 -0.340420  0.384958 -0.104718   
breathes        0.080751  0.441149  0.254588 -0.382777  0.423527  0.286039   
venomous        0.169083 -0.104245 -0.145739  0.098689 -0.242449  0.008528   
fins           -0.037985 -0.280313 -0.223541  0.164796 -0.156328 -0.251157   
legs           -0.087070  0.394009 -0.206686 -0.224918  0.214196  0.043712   
tail            0.092462  0.048973  0.292569 -0.221090  0.210026  0.009482   
domestic       -0.206519  0.207208  0.031586 -0.155610  0.163928  0.063274   
catsize        -0.027243  0.455020 -0.135934 -0.514650  0.574906 -0.349768   
type            0.080012 -0.562384 -0.197520  0.661825 -0.723683  0.022677   

              aquatic  predator   toothed  backbone  breathes  venomous  \
animal name  0.023475 -0.027931 -0.130423 -0.058669  0.080751  0.169083   
hair        -0.473554 -0.154769  0.492531  0.191681  0.441149 -0.104245   
feathers    -0.058552 -0.104430 -0.613631  0.231403  0.254588 -0.145739   
eggs         0.376244  0.011605 -0.642150 -0.340420 -0.382777  0.098689   
milk        -0.362613 -0.029721  0.628168  0.384958  0.423527 -0.242449   
airborne    -0.172638 -0.295181 -0.594311 -0.104718  0.286039  0.008528   
aquatic      1.000000  0.375978  0.053150  0.022463 -0.637506  0.087915   
predator     0.375978  1.000000  0.129452  0.051022 -0.262931  0.115391   
toothed      0.053150  0.129452  1.000000  0.575085 -0.065690 -0.062344   
backbone     0.022463  0.051022  0.575085  1.000000  0.207666 -0.246611   
breathes    -0.637506 -0.262931 -0.065690  0.207666  1.000000 -0.120752   
venomous     0.087915  0.115391 -0.062344 -0.246611 -0.120752  1.000000   
fins         0.604492  0.190302  0.364292  0.209499 -0.617219 -0.033956   
legs        -0.360638 -0.099723 -0.193476 -0.432856  0.369868  0.022964   
tail        -0.034642  0.018947  0.310368  0.731762  0.088952 -0.162724   
domestic    -0.224308 -0.309794  0.069430  0.101733  0.124068 -0.003252   
catsize     -0.111866  0.144790  0.344010  0.356976  0.204125 -0.183748   
type         0.326639  0.061179 -0.471527 -0.828845 -0.519308  0.321476   

                 fins      legs      tail  domestic   catsize      type  
animal name -0.037985 -0.087070  0.092462 -0.206519 -0.027243  0.080012  
hair        -0.280313  0.394009  0.048973  0.207208  0.455020 -0.562384  
feathers    -0.223541 -0.206686  0.292569  0.031586 -0.135934 -0.197520  
eggs         0.164796 -0.224918 -0.221090 -0.155610 -0.514650  0.661825  
milk        -0.156328  0.214196  0.210026  0.163928  0.574906 -0.723683  
airborne    -0.251157  0.043712  0.009482  0.063274 -0.349768  0.022677  
aquatic      0.604492 -0.360638 -0.034642 -0.224308 -0.111866  0.326639  
predator     0.190302 -0.099723  0.018947 -0.309794  0.144790  0.061179  
toothed      0.364292 -0.193476  0.310368  0.069430  0.344010 -0.471527  
backbone     0.209499 -0.432856  0.731762  0.101733  0.356976 -0.828845  
breathes    -0.617219  0.369868  0.088952  0.124068  0.204125 -0.519308  
venomous    -0.033956  0.022964 -0.162724 -0.003252 -0.183748  0.321476  
fins         1.000000 -0.605652  0.204349 -0.093887  0.031705  0.099430  
legs        -0.605652  1.000000 -0.348295  0.073931  0.068791  0.131

In [19]:
df.describe().T

count       mean        std  min   25%   50%   75%   max
animal name  101.0  49.257426  28.968829  0.0  25.0  49.0  74.0  99.0
hair         101.0   0.425743   0.496921  0.0   0.0   0.0   1.0   1.0
feathers     101.0   0.198020   0.400495  0.0   0.0   0.0   0.0   1.0
eggs         101.0   0.584158   0.495325  0.0   0.0   1.0   1.0   1.0
milk         101.0   0.405941   0.493522  0.0   0.0   0.0   1.0   1.0
airborne     101.0   0.237624   0.427750  0.0   0.0   0.0   0.0   1.0
aquatic      101.0   0.356436   0.481335  0.0   0.0   0.0   1.0   1.0
predator     101.0   0.554455   0.499505  0.0   0.0   1.0   1.0   1.0
toothed      101.0   0.603960   0.491512  0.0   0.0   1.0   1.0   1.0
backbone     101.0   0.821782   0.384605  0.0   1.0   1.0   1.0   1.0
breathes     101.0   0.792079   0.407844  0.0   1.0   1.0   1.0   1.0
venomous     101.0   0.079208   0.271410  0.0   0.0   0.0   0.0   1.0
fins         101.0   0.168317   0.376013  0.0   0.0   0.0   0.0   1.0
legs         101.0   2.841584   2.033385  0.0   2.0   4.0   4.0   8.0
tail         101.0   0.742574   0.439397  0.0   0.0   1.0   1.0   1.0
domestic     101.0   0.128713   0.336552  0.0   0.0   0.0   0.0   1.0
catsize      101.0   0.435644   0.498314  0.0   0.0   0.0   1.0   1.0
type         101.0   2.831683   2.102709  1.0   1.0   2.0   4.0   7.0

In [4]:
df['type'].value_counts()

1    41
2    20
4    13
7    10
6     8
3     5
5     4
Name: type, dtype: int64

In [5]:
from sklearn.preprocessing import LabelEncoder
LE = LabelEncoder()
df['animal name'] = LE.fit_transform(df['animal name'])
df

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0              0     1         0     0     1         0        0         1   
1              1     1         0     0     1         0        0         0   
2              2     0         0     1     0         0        1         1   
3              3     1         0     0     1         0        0         1   
4              4     1         0     0     1         0        0         1   
..           ...   ...       ...   ...   ...       ...      ...       ...   
96            95     1         0     0     1         0        0         0   
97            96     1         0     1     0         1        0         0   
98            97     1         0     0     1         0        0         1   
99            98     0         0     1     0         0        0         0   
100           99     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [6]:
X = df.iloc[:,0:17]
#(or)X = df.drop('type',axis=1)
Y = df['type']

In [7]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.3)
X_train.shape,X_test.shape,Y_train.shape,Y_test.shape

((70, 17), (31, 17), (70,), (31,))

In [8]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5,p=2)#p-->2 for eucleidian distance
knn.fit(X_train,Y_train)

KNeighborsClassifier()

In [9]:
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)

In [10]:
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.5285714285714286
Test score is 0.3548387096774194


In [11]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5,p=1)#p-->1 for manhattan distance
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7571428571428571
Test score is 0.5806451612903226


In [12]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=7,p=2)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.5428571428571428
Test score is 0.3870967741935484


In [13]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=7,p=1)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7285714285714285
Test score is 0.6129032258064516


In [14]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=10,p=2)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.45714285714285713
Test score is 0.3548387096774194


In [15]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=10,p=1)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.5857142857142857
Test score is 0.3870967741935484


In [16]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=4,p=2)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.5714285714285714
Test score is 0.41935483870967744


In [17]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=4,p=1)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7857142857142857
Test score is 0.6129032258064516


In [82]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=2,p=2)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7428571428571429
Test score is 0.3548387096774194


In [83]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=2,p=1)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.8571428571428571
Test score is 0.45161290322580644


In [84]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1,p=2)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 1.0
Test score is 0.41935483870967744


In [85]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1,p=1)
knn.fit(X_train,Y_train)
Y_pred_train = knn.predict(X_train)
Y_pred_test = knn.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 1.0
Test score is 0.6451612903225806


# As we see from the above data we conclude that by increasing n_neighbors the Accuracy Score will decrease and by using
# Manhattan distance we are getting high Accuracy Score than Eucledian distance and by decreasing n_neighbors the Accuracy Score 
# will get increasing.

# Q2).Prepare a model for glass classification using KNN

# Data Description:

# RI : refractive index

# Na: Sodium (unit measurement: weight percent in corresponding oxide, as are attributes 4-10)

# Mg: Magnesium

# AI: Aluminum

# Si: Silicon

# K:Potassium

# Ca: Calcium

# Ba: Barium

# Fe: Iron

# Type: Type of glass: (class attribute)
# 1 -- building_windows_float_processed
# 2 --building_windows_non_float_processed
# 3 --vehicle_windows_float_processed
# 4 --vehicle_windows_non_float_processed (none in this database)
# 5 --containers
# 6 --tableware
# 7 --headlamps

In [3]:
import pandas as pd
df2 = pd.read_csv('glass.csv')
df2.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [87]:
df2.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [88]:
df2.corr()

RI        Na        Mg        Al        Si         K        Ca  \
RI    1.000000 -0.191885 -0.122274 -0.407326 -0.542052 -0.289833  0.810403   
Na   -0.191885  1.000000 -0.273732  0.156794 -0.069809 -0.266087 -0.275442   
Mg   -0.122274 -0.273732  1.000000 -0.481799 -0.165927  0.005396 -0.443750   
Al   -0.407326  0.156794 -0.481799  1.000000 -0.005524  0.325958 -0.259592   
Si   -0.542052 -0.069809 -0.165927 -0.005524  1.000000 -0.193331 -0.208732   
K    -0.289833 -0.266087  0.005396  0.325958 -0.193331  1.000000 -0.317836   
Ca    0.810403 -0.275442 -0.443750 -0.259592 -0.208732 -0.317836  1.000000   
Ba   -0.000386  0.326603 -0.492262  0.479404 -0.102151 -0.042618 -0.112841   
Fe    0.143010 -0.241346  0.083060 -0.074402 -0.094201 -0.007719  0.124968   
Type -0.164237  0.502898 -0.744993  0.598829  0.151565 -0.010054  0.000952   

            Ba        Fe      Type  
RI   -0.000386  0.143010 -0.164237  
Na    0.326603 -0.241346  0.502898  
Mg   -0.492262  0.083060 -0.744993  
Al    0.479404 -0.074402  0.598829  
Si   -0.102151 -0.094201  0.151565  
K    -0.042618 -0.007719 -0.010054  
Ca   -0.112841  0.124968  0.000952  
Ba    1.000000 -0.058692  0.575161  
Fe   -0.058692  1.000000 -0.188278  
Type  0.575161 -0.188278  1.000000

In [92]:
df2['Type'].value_counts()

2    76
1    70
7    29
3    17
5    13
6     9
Name: Type, dtype: int64

In [94]:
X = df2.iloc[:,0:9]
Y = df2['Type']

In [98]:
from sklearn.preprocessing import MinMaxScaler
mm = MinMaxScaler()
mm_x = mm.fit_transform(X)
mm_x = pd.DataFrame(mm_x)
mm_x.columns = list(X)
mm_x

RI        Na        Mg        Al        Si         K        Ca  \
0    0.432836  0.437594  1.000000  0.252336  0.351786  0.009662  0.308550   
1    0.283582  0.475188  0.801782  0.333333  0.521429  0.077295  0.223048   
2    0.220808  0.421053  0.790646  0.389408  0.567857  0.062802  0.218401   
3    0.285777  0.372932  0.821826  0.311526  0.500000  0.091787  0.259294   
4    0.275241  0.381955  0.806236  0.295950  0.583929  0.088567  0.245353   
..        ...       ...       ...       ...       ...       ...       ...   
209  0.223003  0.512782  0.000000  0.806854  0.500000  0.012882  0.348513   
210  0.250219  0.630075  0.000000  0.529595  0.580357  0.000000  0.276022   
211  0.417032  0.545865  0.000000  0.538941  0.644643  0.000000  0.279740   
212  0.235294  0.548872  0.000000  0.514019  0.678571  0.000000  0.283457   
213  0.261633  0.526316  0.000000  0.557632  0.633929  0.000000  0.296468   

           Ba   Fe  
0    0.000000  0.0  
1    0.000000  0.0  
2    0.000000  0.0  
3    0.000000  0.0  
4    0.000000  0.0  
..        ...  ...  
209  0.336508  0.0  
210  0.504762  0.0  
211  0.520635  0.0  
212  0.498413  0.0  
213  0.530159  0.0  

[214 rows x 9 columns]

In [99]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.3)
X_train.shape,X_test.shape,Y_train.shape,Y_test.shape

((149, 9), (65, 9), (149,), (65,))

In [100]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=5,p=2)
knc.fit(X_train,Y_train)

KNeighborsClassifier()

In [101]:
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)

In [102]:
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7248322147651006
Test score is 0.6461538461538462


In [103]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=5,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7516778523489933
Test score is 0.6615384615384615


In [104]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=6,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7248322147651006
Test score is 0.6307692307692307


In [107]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=6,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7315436241610739
Test score is 0.6461538461538462


In [108]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=8,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.697986577181208
Test score is 0.6307692307692307


In [109]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=8,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.6845637583892618
Test score is 0.676923076923077


In [110]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=10,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.6577181208053692
Test score is 0.6461538461538462


In [111]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=10,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.6778523489932886
Test score is 0.676923076923077


In [112]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=4,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.7449664429530202
Test score is 0.6461538461538462


In [113]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=4,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.8053691275167785
Test score is 0.6923076923076923


In [114]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=2,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.8187919463087249
Test score is 0.676923076923077


In [115]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=2,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 0.8120805369127517
Test score is 0.6923076923076923


In [116]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=1,p=2)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 1.0
Test score is 0.7384615384615385


In [117]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier(n_neighbors=1,p=1)
knc.fit(X_train,Y_train)
Y_pred_train = knc.predict(X_train)
Y_pred_test = knc.predict(X_test)
from sklearn.metrics import accuracy_score
print('Training score is',accuracy_score(Y_train,Y_pred_train))
print('Test score is',accuracy_score(Y_test,Y_pred_test))

Training score is 1.0
Test score is 0.7076923076923077


# For normal data the accuracy score is high for n_neighbors = 3 and p =2 and p = 1 the accuracy score is 0.83 and 
# for n_neighbors = 5 and for p = 2 the train accuracy score is 0.73 and test accuracy score is 0.65 and for p=1 the train 
# accuracuy score is 0.72 and test accuracy score is 0.65, As the n_neighbors is high and acccuracy score is low